src/ir/module-utils.cpp

```diff
@@ -75,6 +75,42 @@ struct CodeScanner
       counts.note(get->ref->type);
     } else if (auto* set = curr->dynCast<StructSet>()) {
       counts.note(set->ref->type);
+    } else if (auto* get = curr->dynCast<LocalGet>()) {
+      // Make sure that local types are all included in the types we know.
+      // Normally they already are: local types appear either in the function
+      // signature (for params) or the function vars, all of which are already
+      // counted. However, we also need to collect heap types *during* a
+      // signature update, when refining the type of a parameter: we first fix
+      // up local.gets and other expressions, then we collect heap types and we
+      // use that to update the signature everywhere it is used, in particular,
+      // in the function itself. We must do that all at once, and so at the
+      // point in time that we collect heap types the local.get has been updated
+      // to match the new param type, but the param type as declared in the
+      // function signature has not yet been updated. If the local.get is the
+      // only place in the entire program that uses this type then we'd run into
+      // an internal error later, so make sure the type is included here.
+      //
+      // This problem is specific to local.get because the type of local.get is
+      // dependent on the function's signature. Therefore local.get types must
+      // be changed atomically with a signature change to one of its params.
+      // However, for modularity we have separate utility code for each and not
+      // a single big monolithic helper that does both, so we need to be a
+      // little flexible here to handle IR that is partially updated,
+      // specifically, has local.get types updated but not the signature yet.
```

Comment:
I took a look at SignatureRefining.cpp, but it wasn't obvious to me how the code modifications interact with heap type collection, so a more detailed walkthrough of the problem would be helpful.

Does the fact that `LUB::getResultsLUB` have side effects play into this? We might want to consider having it return a list of Expressions that need to be updated rather than having it perform the updates itself to keep that parallel analysis read-only.

Reply:
Sorry, yeah, it's a little complicated, both what the underlying issue is but also how the fuzz testcase actually manages to cause the underlying issue to hit an error.

The key issue is that `local.get` types are tied to the function signature. That means that we need to change them atomically when we change the signature. More specifically, we update `local.get` in `TypeUpdating::updateParamTypes()`, which if we are changing function signatures means we do a "partial update" - the IR will only be in a consistent state after we also change the signatures, which `GlobalTypeRewriter::updateSignatures` does later. So another approach here would be to avoid a partial update and leave more to that later step. That is more code, but might be cleaner - I pushed that now.

Summary of current state of the code in this PR:

Optionally avoid updating types in `TypeUpdating::updateParamTypes()`. That update is incomplete if the function signature is also changing, which is the case in SignatureRefining (but not DeadArgumentElimination). "Incomplete" means that we updated the `local.get` type, but the function signature does not match yet. That incomplete state can hit an internal error in `GlobalTypeRewriter::updateSignatures` where it updates types. To avoid that, do the entire full update only there (in `GlobalTypeRewriter::updateSignatures`).

 (diff is smaller without whitespace).

> Does the fact that LUB::getResultsLUB have side effects play into this? 

The refinalize there helps the fuzz bug hit the error, but it's not the key issue (changes there would just be a workaround, in other words).

Reply:
Friendly ping on this PR.